psi4/driver/procrouting/scf/scf_iterator.py

```diff
@@ -0,0 +1,583 @@
+    # Properties
+    #  Comments so that autodoc utility will find these PSI variables
+    #  Process::environment.globals["SCF DIPOLE X"] =
+    #  Process::environment.globals["SCF DIPOLE Y"] =
+    #  Process::environment.globals["SCF DIPOLE Z"] =
+    #  Process::environment.globals["SCF QUADRUPOLE XX"] =
+    #  Process::environment.globals["SCF QUADRUPOLE XY"] =
+    #  Process::environment.globals["SCF QUADRUPOLE XZ"] =
+    #  Process::environment.globals["SCF QUADRUPOLE YY"] =
+    #  Process::environment.globals["SCF QUADRUPOLE YZ"] =
+    #  Process::environment.globals["SCF QUADRUPOLE ZZ"] =
+
+    # Orbitals are always saved, in case an MO guess is requested later
+    # save_orbitals()
+
+    self.finalize()
+
+    core.print_out("\nComputation Completed\n")
+
+    return energy
+
+
+core.HF.initialize = scf_initialize
+core.HF.iterations = scf_iterate
+core.HF.compute_energy = scf_compute_energy
+core.HF.finalize_energy = scf_finalize_energy
+
+
+def _validate_damping():
+    # Q: I changed the enabled criterion get_option <-- has_option_changed
+    enabled = (core.get_option('SCF', 'DAMPING_PERCENTAGE') > 0.0)
+    if enabled:
+        parameter = core.get_option('SCF', "DAMPING_PERCENTAGE")
+        if parameter < 0.0 or parameter > 100.0:
+            raise ValidationError('SCF DAMPING_PERCENTAGE ({}) must be between 0 and 100'.format(parameter))
+
+        stop = core.get_option('SCF', 'DAMPING_CONVERGENCE')
+        if stop < 0.0:
+            raise ValidationError('SCF DAMPING_CONVERGENCE ({}) must be > 0'.format(stop))
+
+    return enabled
+
+
+def _validate_diis():
+    enabled = bool(core.get_option('SCF', 'DIIS'))
+    if enabled:
+        start = core.get_option('SCF', 'DIIS_START')
+        if start < 1:
+            raise ValidationError('SCF DIIS_START ({}) must be at least 1'.format(start))
+
+        minvecs = core.get_option('SCF', 'DIIS_MIN_VECS')
+        if minvecs < 1:
+            raise ValidationError('SCF DIIS_MIN_VECS ({}) must be at least 1'.format(minvecs))
+
+        maxvecs = core.get_option('SCF', 'DIIS_MAX_VECS')
+        if maxvecs < minvecs:
+            raise ValidationError('SCF DIIS_MAX_VECS ({}) must be at least DIIS_MIN_VECS ({})'.format(
+                                  maxvecs, minvecs))
+
+    return enabled
+
+
+def _validate_frac():
+    enabled = (core.get_option('SCF', 'FRAC_START') != 0)
+    if enabled:
+        if enabled < 0:
+            raise ValidationError('SCF FRAC_START ({}) must be at least 1'.format(enabled))
+
+    return enabled
+
+
+def _validate_MOM():
+    enabled = (core.get_option('SCF', "MOM_START") != 0 and len(core.get_option('SCF', "MOM_OCC")) > 0)
+    if enabled:
+        start = core.get_option('SCF', "MOM_START")
+        if enabled < 0:
+            raise ValidationError('SCF MOM_START ({}) must be at least 1'.format(start))
+
+    # note that this is excited-MOM-enabled, which is the more interesting flag
+    return enabled
+
+
+def _validate_soscf():
```

Comment:
Hate to ask, but can you doc some of the logic trees on these?

Reply:
I didn't change any of the logic on damp/diis/soscf/mom/frac (I hope). Doc for review, doc for general interest in sphinxman/hf.rst or read_options.cc, doc for programmers meeting the SCF code?

The checks in the `def _validate_*` fns are pretty much just sanity checks that may/may not have been elsewhere.

Reply:
Mostly just a doc for someone reading the code, probably something as simple as "Validates SOSCF start, convergences, and iteration options" would work just fine.